design/sketch.webidl

```diff
@@ -387,23 +388,23 @@ enum GPUInputStepMode {
 };
 
 dictionary GPUVertexAttributeDescriptor {
-    u32 shaderLocation;
-    u32 inputSlot;
     u32 offset;
     GPUVertexFormat format;
+    GPUShaderAttributeIndex attributeIndex;
 };
 
-dictionary GPUVertexInputDescriptor {
-    u32 inputSlot;
+dictionary GPUVertexBufferDescriptor {
     u32 stride;
     GPUInputStepMode stepMode;
+    sequence<GPUVertexAttributeDescriptor> attributes;
 };
 
-dictionary GPUInputStateDescriptor {
+dictionary GPUVertexBufferStateDescriptor {
     GPUIndexFormat indexFormat;
 
-    sequence<GPUVertexAttributeDescriptor> attributes;
-    sequence<GPUVertexInputDescriptor> inputs;
+    // This object has keys of 0..15 and values of type GPUVertexBufferDescriptor.
+    // (It's not possible in WebIDL to describe a dictionary with keys "0" through "15".)
+    object vertexBuffers;
```

Comment:
@heycam Is this really not possible?

Reply:
Yes, this is not possible. (There is an escaping prefix `_` for identifiers but the characters following must still look like a normal identifier, and this is only used for escaping existing keywords.)

I think it's odd for a Web API to have array-index-like members like this.  What's the downside of using `sequence<GPUVertexInputDescriptor>` here?

Reply:
Please take a look at the [alternatives doc](https://github.com/kainino0x/gpuweb/pull/2/files).

If by `sequence<GPUVertexInputDescriptor>` you mean WITH the index parameter, that is option 2. The additional indirection was opposed by some in favor of the relatively more straightforward-looking keys-as-indices.

If you mean WITHOUT the index parameter, that's option 3 and the sequence must be sparse - that is, `sequence<GPUVertexBufferDescriptor?>` (since afaik WebIDL doesn't really have sparse arrays).

Reply:
All that said, I'm currently not that strongly in favor of any of these proposals over any of the others. It is up for discussion.

Reply:
Thanks for that pointer. Do you think authors will tend to pass an Array object (with holes or explicit `undefined` values) or a plain object with index-like keys, if the IDL used `object`?  One of the points made about option 3 is that it obscures the index of each descriptor but even with the `object` and in-prose handling of reading things off the object, the author could still pass an Array object and have it work. Unless the prose explicitly checked for an Array and threw an exception I suppose.

If requiring the index to be specified is not a hard requirement, then I think using `sequence<GPUVertexBufferDescriptor?>` is the most straightforward thing.

Another option would be to use a dictionary with names that aren't array-index-like, e.g.

```
dictionary GPUVertexInputBuffers {
    optional GPUVertexBufferDescriptor v0;
    optional GPUVertexBufferDescriptor v1;
    ...
};
```

Reply:
Thanks for the input. I envisioned primarily using object syntax and passing raw objects to the API. We could allow or disallow arrays.

Your last suggestion looks like Option 5 (the last one in the PR I linked). It is definitely possible, but I think applications will sometimes want to e.g. have a `for` loop over slots 2-5, which would be harder and slower with the valid identifier names.

Reply:
> If you mean WITHOUT the index parameter, that's option 3 and the sequence must be sparse - that is, sequence<GPUVertexBufferDescriptor?> (since afaik WebIDL doesn't really have sparse arrays).

There is no need for it to be sparse, no need for `?`. If no vertex attributes are using a buffer at some index, the corresponding `GPUVertexBufferDescriptor` would just have an empty list of attributes.

`object vertexBuffers;` doesn't look appealing at all to me